src/borg/_hashindex.c

```diff
@@ -731,23 +731,23 @@ static int
 hashindex_set(HashIndex *index, const unsigned char *key, const void *value)
 {
     int start_idx;
-    int idx = hashindex_lookup(index, key, &start_idx);
+    int idx = hashindex_lookup(index, key, &start_idx);  /* if idx < 0: start_idx -> EMPTY or DELETED */
     uint8_t *ptr;
     if(idx < 0)
     {
         if(index->num_entries > index->upper_limit) {
+            /* hashtable too full, grow it! */
             if(!hashindex_resize(index, grow_size(index->num_buckets))) {
                 return 0;
             }
-            start_idx = hashindex_index(index, key);
+            /* we have just built a fresh hashtable and removed all tombstones,
+             * so we only have EMPTY or USED buckets, but no DELETED ones any more.
+             */
+            idx = hashindex_lookup(index, key, &start_idx);
+            assert(idx < 0);
+            assert(BUCKET_IS_EMPTY(index, start_idx));
         }
         idx = start_idx;
-        while(!BUCKET_IS_EMPTY(index, idx) && !BUCKET_IS_DELETED(index, idx)) {
-            idx++;
-            if (idx >= index->num_buckets){
-                idx -= index->num_buckets;
-            }
-        }
         if(BUCKET_IS_EMPTY(index, idx)){
             index->num_empty--;
             if(index->num_empty < index->min_empty) {
```

Comment:
I think it would be logically cleaner to move the line `index->num_empty--` to later in this `if` block, so that we aren't calling `hashindex_resize` and `hashindex_lookup` on a hash table with incorrect metadata.  The line `if(index->num_empty < index->min_empty)` would then change to `<=`.

Reply:
This actually fixes a bug from the old code.

Old code:
Considering we will use up one empty slot in below code, we first decremented `num_empty`.
Then, hashindex_resize() internally **overwrote** num_empty (recomputing it from `num_buckets - num_entries`). Because `num_entries` is not incremented yet, that resulted in a wrong value (we lost the decrement operation).

Reply:
Nice observation!  I knew something was fishy about that code.  I doubt it leads to any data loss, though.

While looking at hashindex_resize, I thought that it would be good to add an assertion:
```
assert(index->num_entries == new->num_entries);
```

Reply:
It only uses the "new" buckets table and points to that from the "old" data structure.
Besides that, it continues to work with the old data structure and just selectively copies some stuff from "new" (but not num_entries).

But yes, we could check if the copying loop results in consistent num_entries values.

Reply:
done.